src/libostree/ostree-sysroot-deploy.c

```diff
@@ -3614,25 +3614,45 @@ ostree_sysroot_deployment_set_kargs_in_place (OstreeSysroot     *self,
   if (!_ostree_sysroot_ensure_writable (self, error))
     return FALSE;
 
-  g_assert (!ostree_deployment_is_staged (deployment));
+  if (ostree_deployment_is_staged (deployment))
+    {
+      if (!_ostree_sysroot_reload_staged (self, error))
+        return FALSE;
 
-  OstreeBootconfigParser *new_bootconfig = ostree_deployment_get_bootconfig (deployment);
-  ostree_bootconfig_parser_set (new_bootconfig, "options", kargs_str);
+      g_autoptr(OstreeKernelArgs) kargs = ostree_kernel_args_from_string (kargs_str);
+      g_auto(GStrv) kargs_strv = ostree_kernel_args_to_strv (kargs);
+      
+      OstreeSysrootDeployTreeOpts opts = {
+        .override_kernel_argv = kargs_strv,
+      };
+      g_autoptr(OstreeDeployment) new_deployment = NULL;
+      const char *osname = ostree_deployment_get_osname (self->staged_deployment);
+      g_autoptr(OstreeDeployment) merge_deployment = ostree_sysroot_get_merge_deployment (self, osname);
+      if (!ostree_sysroot_stage_tree_with_options(self, osname, ostree_deployment_get_csum (self->staged_deployment),
```

Comment:
This will replace the staged deployment instead of editing it in-place. So we'd possibly lose other attributes of the staged deployment in the process.

To elaborate a bit more on https://github.com/ostreedev/ostree/issues/2679#issuecomment-1194483321, we need to do part of what `_ostree_sysroot_reload_staged` does ourselves, which is read the contents of `_OSTREE_SYSROOT_RUNSTATE_STAGED` into a `GVariantDict`. Then we can modify the `kargs` member as necessary, and then rewrite `_OSTREE_SYSROOT_RUNSTATE_STAGED` with our modifications.

Reply:
Thanks @jlebon!
So I should write another function like `_ostree_sysroot_reload_staged`, only modify `kargs`, to rewrite 
 `_OSTREE_SYSROOT_RUNSTATE_STAGED` like https://github.com/ostreedev/ostree/blob/main/src/libostree/ostree-sysroot-deploy.c#L3355?

Reply:
Yes, exactly! You can see if there's something worth factoring out of those functions so we can use them here too. They might be too thin to be worth it.